[builder error] match + &

### Describe the bug

```
==================
C:/Users/russo/AppData/Local/Temp/v_0/bug1.01J2KZSNTGBA77PHH8BD7V62CY.tmp.c:6979: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/russo/AppData/Local/Temp/v_0/bug1.01J2KZSNTGBA77PHH8BD7V62CY.tmp.c:13360: error: lvalue expected
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .

C:\Users\russo\Desktop\vbugs>v run bug1.v
==================
C:/Users/russo/AppData/Local/Temp/v_0/bug1.01J2KZWH3Z9FQ7DWEPSBXQMSBC.tmp.c:6979: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/russo/AppData/Local/Temp/v_0/bug1.01J2KZWH3Z9FQ7DWEPSBXQMSBC.tmp.c:13360: error: lvalue expected
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Reproduction Steps

```v
module main

pub struct Expr1 {}
pub struct Expr2 {}

pub type Expr = Expr1 | Expr2

pub struct Gen {
}

fn (mut g Gen) foo(mut expr &Expr1) {
}

pub fn (mut g Gen) bar(mut expr &Expr) {
	match mut expr {
		Expr1 { g.foo(mut &expr) } // <--- &
		Expr2 {}
	}
}

fn main() {
}
```

### Expected Behavior

ok

### Current Behavior

builder error

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.6 7908957

### Environment details (OS name and version, etc.)

V full version: V 0.4.6 58a8fc6.7908957
OS: windows,  Windows 10 Pro v19042 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: M:\_projects_skyrim\papyrus-compiler
vexe: M:\_dev\vlang\v.exe
vexe mtime: 2024-07-12 16:28:32

vroot: OK, value: M:\_dev\vlang
VMODULES: OK, value: C:\Users\russo\.vmodules
VTMP: OK, value: C:\Users\russo\AppData\Local\Temp\v_0

Git version: git version 2.45.1.windows.1
Git vroot status: weekly.2024.20-247-g79089578
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc: N/A
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2608,8 +2608,14 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.write('&/*mut*/')
 	} else if arg.is_mut && arg_typ.is_ptr() && expected_type.is_ptr()
 		&& g.table.sym(arg_typ).kind == .struct_ && expected_type == arg_typ.ref() {
-		g.write('&/*mut*/')
-		g.expr(arg.expr)
+		if arg.expr is ast.PrefixExpr && arg.expr.op == .amp {
+			g.arg_no_auto_deref = true
+			g.expr(arg.expr)
+			g.arg_no_auto_deref = false
+		} else {
+			g.write('&/*mut*/')
+			g.expr(arg.expr)
+		}
 		return
 	} else if exp_is_ptr && !arg_is_ptr && !(arg_sym.kind == .alias
 		&& g.table.unaliased_type(arg_typ).is_pointer() && expected_type.is_pointer()) {

--- a/vlib/v/tests/fn_mut_arg_of_sumtype_ref_test.v
+++ b/vlib/v/tests/fn_mut_arg_of_sumtype_ref_test.v
@@ -1,0 +1,29 @@
+module main
+
+pub struct Expr1 {}
+
+pub struct Expr2 {}
+
+pub type Expr = Expr1 | Expr2
+
+pub struct Gen {
+}
+
+fn (mut g Gen) foo(mut expr &Expr1) string {
+	return '${expr}'
+}
+
+pub fn (mut g Gen) bar(mut expr Expr) string {
+	return match mut expr {
+		Expr1 { g.foo(mut &expr) }
+		Expr2 { '' }
+	}
+}
+
+fn test_fn_mut_arg_of_sumtype_ref() {
+	mut expr := &Expr1{}
+	mut g := Gen{}
+	ret := g.bar(mut expr)
+	println(ret)
+	assert ret == 'Expr1{}'
+}

